File: packages/sillytavern-compat/src/generation.ts
import type { JsonObject, JsonValue } from "@zhuangsheng/api-client";
import { integer, number, object, string } from "./json";
import { warn, type ImportParts } from "./support";

const OPENAI_EXTRA = ["frequency_penalty", "presence_penalty", "top_k", "top_a", "min_p", "repetition_penalty"];
const COMPLETION_EXTRA = ["top_k", "min_p", "rep_pen", "rep_pen_range", "typical_p", "tfs"];

export function importOpenAiGeneration(document: JsonObject, parts: ImportParts) {
  setGeneration(parts, {
    temperature: number(document.temperature), topP: number(document.top_p),
    maxOutputTokens: positiveInteger(document.openai_max_tokens), stop: stopStrings(document),
    seed: nonNegativeInteger(document.seed),
  });
  setExtensions(document, OPENAI_EXTRA, parts, "non-portable generation fields were preserved as OpenAI extra-body options");
}

export function importCompletionGeneration(document: JsonObject, parts: ImportParts) {
  setGeneration(parts, {
    temperature: number(document.temp), topP: number(document.top_p),
    maxOutputTokens: positiveInteger(document.genamt) ?? positiveInteger(document.max_length),
    stop: stopStrings(document), seed: nonNegativeInteger(document.seed),
  });
  setExtensions(document, COMPLETION_EXTRA, parts, "text-completion sampler fields were preserved as provider extra-body options");
}

function setGeneration(parts: ImportParts, value: JsonObject) {
  if (value.temperature !== null || value.topP !== null || value.maxOutputTokens !== null || value.seed !== null || (value.stop as JsonValue[]).length) {
    parts.generation = value;
  }
}

function setExtensions(document: JsonObject, keys: string[], parts: ImportParts, message: string) {
  const extraBody = Object.fromEntries(keys.flatMap((key) => {
    const value = document[key];
    return value !== undefined && (typeof value === "boolean" || number(value) !== null) ? [[key, value]] : [];
  })) as JsonObject;
  if (!Object.keys(extraBody).length) return;
  parts.providerExtensions = { openai: { options: {}, extraBody, extraHeaders: {} }, claude: null, gemini: null };
  warn(parts, "sillytavern_provider_extensions", message);
}

function stopStrings(document: JsonObject): JsonValue[] {
  for (const key of ["stop", "stopping_strings"]) {
    const values = document[key];
    if (Array.isArray(values)) return values.filter((value): value is string => typeof value === "string");
  }
  const raw = string(document.custom_stopping_strings);
  if (!raw) return [];
  try { const value = JSON.parse(raw); return Array.isArray(value) ? value.filter((item): item is string => typeof item === "string") : []; }
  catch { return []; }
}

function positiveInteger(value: JsonValue | undefined) { const result = integer(value); return result !== null && result > 0 ? result : null; }
function nonNegativeInteger(value: JsonValue | undefined) { const result = integer(value); return result !== null && result >= 0 ? result : null; }

File: crates/core/src/compatibility/sillytavern/preset/generation.rs
use std::collections::BTreeMap;

use serde_json::Value;

use crate::graph::{GenerationOptionsIr, ProviderExtensionsIr, ProviderExtraIr};

use super::support::{ImportParts, warning};

pub(super) fn import_openai_generation(document: &Value, parts: &mut ImportParts) {
    let generation = GenerationOptionsIr {
        temperature: number(document, "temperature"),
        top_p: number(document, "top_p"),
        max_output_tokens: unsigned(document, "openai_max_tokens"),
        stop: stop_strings(document),
        seed: document
            .get("seed")
            .and_then(Value::as_i64)
            .filter(|value| *value >= 0),
    };
    if generation != empty_generation() {
        parts.generation = Some(generation);
    }
    let extra_body = provider_options(
        document,
        &[
            "frequency_penalty",
            "presence_penalty",
            "top_k",
            "top_a",
            "min_p",
            "repetition_penalty",
        ],
    );
    if !extra_body.is_empty() {
        parts.provider_extensions = Some(ProviderExtensionsIr {
            openai: Some(ProviderExtraIr {
                options: BTreeMap::new(),
                extra_body,
                extra_headers: BTreeMap::new(),
            }),
            ..Default::default()
        });
        warning(
            &mut parts.warnings,
            "sillytavern_provider_extensions",
            "non-portable generation fields were preserved as OpenAI extra-body options",
            None,
        );
    }
}

pub(super) fn import_text_completion_generation(document: &Value, parts: &mut ImportParts) {
    let generation = GenerationOptionsIr {
        temperature: number(document, "temp"),
        top_p: number(document, "top_p"),
        max_output_tokens: unsigned(document, "genamt")
            .or_else(|| unsigned(document, "max_length")),
        stop: stop_strings(document),
        seed: document
            .get("seed")
            .and_then(Value::as_i64)
            .filter(|value| *value >= 0),
    };
    if generation != empty_generation() {
        parts.generation = Some(generation);
    }
    let extra_body = provider_options(
        document,
        &[
            "top_k",
            "min_p",
            "rep_pen",
            "rep_pen_range",
            "typical_p",
            "tfs",
        ],
    );
    if !extra_body.is_empty() {
        parts.provider_extensions = Some(ProviderExtensionsIr {
            openai: Some(ProviderExtraIr {
                options: BTreeMap::new(),
                extra_body,
                extra_headers: BTreeMap::new(),
            }),
            ..Default::default()
        });
        warning(
            &mut parts.warnings,
            "sillytavern_provider_extensions",
            "text-completion sampler fields were preserved as provider extra-body options",
            None,
        );
    }
}

fn empty_generation() -> GenerationOptionsIr {
    GenerationOptionsIr {
        temperature: None,
        top_p: None,
        max_output_tokens: None,
        stop: Vec::new(),
        seed: None,
    }
}

fn number(document: &Value, key: &str) -> Option<f64> {
    document
        .get(key)
        .and_then(Value::as_f64)
        .filter(|value| value.is_finite())
}

fn unsigned(document: &Value, key: &str) -> Option<u64> {
    document
        .get(key)
        .and_then(Value::as_u64)
        .filter(|value| *value > 0)
}

fn provider_options(document: &Value, keys: &[&str]) -> BTreeMap<String, Value> {
    keys.iter()
        .filter_map(|key| {
            document
                .get(*key)
                .filter(|value| value.is_number() || value.is_boolean())
                .map(|value| ((*key).to_owned(), value.clone()))
        })
        .collect()
}

fn stop_strings(document: &Value) -> Vec<String> {
    for key in ["stop", "stopping_strings"] {
        if let Some(values) = document.get(key).and_then(Value::as_array) {
            return values
                .iter()
                .filter_map(Value::as_str)
                .map(str::to_owned)
                .collect();
        }
    }
    let Some(raw) = document
        .get("custom_stopping_strings")
        .and_then(Value::as_str)
    else {
        return Vec::new();
    };
    serde_json::from_str::<Vec<String>>(raw).unwrap_or_default()
}
